### PR TITLE
EventStore

### PR DESCRIPTION
**Pull request by davidcorcoran**
A DB backed EventStore with methods exposed over HTTP. See the README.md for more info.

### CODE REVIEW

File: base/src/freetrm/eventstore/DeDuplicator.scala [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import scala.util.{Failure, Success}
+import scala.util.control.NonFatal
+
+
+class DeDuplicator(reader: EventSourceReader, writer: EventSourceWriter)
+                  (implicit system: ActorSystem) extends EventSourceWriter with Log {
+  
+  private val dedup = new ConcurrentHashMap[String, FutureTask[Future[SingleTopicDeDuplicator]]]
```
Comment (cage433) [inline base/src/freetrm/eventstore/DeDuplicator.scala]: Would a `Topic` tiny type make this more obvious?

File: base/src/freetrm/eventstore/DeDuplicator.scala [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+              dedup.update(key, contentHash, version)
+            case _ =>
+          }.runWith(Sink.ignore).map {
+            case _ => dedup
+          }
+          
+        case None => 
+          Future.successful(new SingleTopicDeDuplicator(topic))
+          
+        case o =>
```
Comment (cage433) [inline base/src/freetrm/eventstore/DeDuplicator.scala]: This case shouldn't be necessary

File: base/src/freetrm/eventstore/DeDuplicator.scala [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  case class WriteResult(wasDuplicate: Boolean, version: EventVersionPair)
+  
+  def contentHash(content: String): String = {
+    val digest = MessageDigest.getInstance("MD5").digest(content.getBytes)
+    digest.map("%02x".format(_)).mkString
+  }
+  
+  case class SingleTopicDeDuplicator(topic: String) {
+
+    private val map = new ConcurrentHashMap[CharSequence, (CharSequence, EventVersionPair)]()
```
Comment (cage433) [inline base/src/freetrm/eventstore/DeDuplicator.scala]: Tiny types and/or a better name than `map`?

File: base/src/freetrm/eventstore/DeDuplicator.scala [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+      map.put(key, (contentHash, version)) match {
+        case (previousValue, previousSeq) if previousValue == contentHash =>
+          log.warn(s"You wrote the same version again for ($key, $contentHash, $version), previous: $previousSeq")
+          version
+        case _ =>
+          version
+      }
+    }
+
+    def size: Int = map.size
```
Comment (cage433) [inline base/src/freetrm/eventstore/DeDuplicator.scala]: Seems an odd public method for this class. It's unused - left over from debugging?

File: base/src/freetrm/eventstore/DeDuplicator.scala [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  
+  case class SingleTopicDeDuplicator(topic: String) {
+
+    private val map = new ConcurrentHashMap[CharSequence, (CharSequence, EventVersionPair)]()
+    
+    def isDuplicate(key: String, contentHash: String): Boolean = {
+      previousVersion(key, contentHash).nonEmpty
+    }
+    
+    def previousVersion(key: String, contentHash: String): Option[EventVersionPair] = {
```
Comment (cage433) [inline base/src/freetrm/eventstore/DeDuplicator.scala:87]: should be private

Comment (cage433) [inline base/src/freetrm/eventstore/DeDuplicator.scala:87]: no it shouldn't - ignore

File: base/src/freetrm/eventstore/DeDuplicator.scala [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  def contentHash(content: String): String = {
+    val digest = MessageDigest.getInstance("MD5").digest(content.getBytes)
+    digest.map("%02x".format(_)).mkString
+  }
+  
+  case class SingleTopicDeDuplicator(topic: String) {
+
+    private val map = new ConcurrentHashMap[CharSequence, (CharSequence, EventVersionPair)]()
+    
+    def isDuplicate(key: String, contentHash: String): Boolean = {
```
Comment (cage433) [inline base/src/freetrm/eventstore/DeDuplicator.scala]: unused

File: base/src/freetrm/eventstore/DeDuplicator.scala [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+          version
+        case _ =>
+          version
+      }
+    }
+
+    def size: Int = map.size
+  }
+
+  def empty(topic: String): SingleTopicDeDuplicator = new SingleTopicDeDuplicator(topic)
```
Comment (cage433) [inline base/src/freetrm/eventstore/DeDuplicator.scala]: should move to tests

File: base/src/freetrm/eventstore/http/EventStoreHttpServer.scala [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+                          (implicit system: ActorSystem)
+  extends Directives with Log {
+
+
+  private val dedup = new DeDuplicator(reader, writer)
+  private implicit val executionContext: ExecutionContextExecutor = system.dispatcher
+  private implicit val mat = ActorMaterializer()
+
+  private def withGzipSupport = {
+    (decodeRequestWith(NoCoding) | decodeRequestWith(Gzip)) &
```
Comment (cage433) [inline base/src/freetrm/eventstore/http/EventStoreHttpServer.scala]: Can drop one of the `&` clauses

File: base/src/freetrm/eventstore/http/EventStoreHttpServer.scala [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    Future {
+      cred match {
+        case Some(BasicHttpCredentials(user, `cookie`)) => Right(user)
+        case _ => Left(HttpChallenge("Basic", "freetrm"))
+      }
+    }
+  }
+
+  implicit def rejectionHandler: RejectionHandler =
+    RejectionHandler.newBuilder()
```
Comment (cage433) [inline base/src/freetrm/eventstore/http/EventStoreHttpServer.scala:56]: Is this path tested?

File: base/src/freetrm/eventstore/http/EventStoreHttpServer.scala [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        case result =>
+          response(OK, EventStoreResponseDuplicate(result.version))
+      },
+      e =>
+        s"Failed to publish to event store: $topic: $event"
+    )
+  }
+
+
+  def consume(topic: String,
```
Comment (cage433) [inline base/src/freetrm/eventstore/http/EventStoreHttpServer.scala]: Should be called `streamEvents`

File: base/src/freetrm/eventstore/http/EventStoreHttpServer.scala [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+      .consumeNotifications
+      .map {
+        e => ServerSentEvent(e.toJson.prettyPrint)
+      }
+      .keepAlive(1.second, () => {
+        ServerSentEvent.heartbeat
+      })
+  }
+
+  def surface(topic: String): Source[ServerSentEvent, NotUsed] = {
```
Comment (cage433) [inline base/src/freetrm/eventstore/http/EventStoreHttpServer.scala]: `newestSurface` perhaps? `surface` is used elsewhere in the system to mean the latest as of a particular version

File: base/src/freetrm/eventstore/http/HttpReader.scala [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    val uri = s"/es/surface/${enc(topic)}"
+    val req: Source[HttpResponse, NotUsed] = getRequest(uri)
+
+    consume[EventSourceEvent](req, response => s"Failed calling 'surface': $response")
+  }
+
+  private def consume[T](source: Source[HttpResponse, NotUsed], failMsg: String => String)
+                        (implicit writer: spray.json.JsonReader[T]) = {
+    source
+      .mapAsync(1) {
```
Comment (cage433) [inline base/src/freetrm/eventstore/http/HttpReader.scala]: reason for `mapAsync(1)` isn't obvious. I'd add a comment that although `mapAsync` guarantees order of future completion, `mapAsync(n > 1)` could cause a successful event to be streamed after a failure, whereas we would want to stop processing and (perhaps) retry.

File: base/src/freetrm/eventstore/http/HttpReader.scala [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+  override def maxSeqNo(topic: String): Future[Option[Long]] = {
+    jsonRPC[Option[Long]](s"/es/maxSeqNo/${enc(topic)}")
+  }
+
+  def close(t: Option[Throwable]) = {
+    log.info("HttpReader.close called")
+  }
+
+  private def getRequest(uri: String) = {
+    require(uri.charAt(0) == '/', s"Invalid URI (needs leading slash): $uri")
```
Comment (cage433) [inline base/src/freetrm/eventstore/http/HttpReader.scala]: `uri.head`?

File: base/src/freetrm/eventstore/http/HttpWriter.scala [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    val settings = ClientConnectionSettings(system)
+    Http().outgoingConnection(host, port, settings = settings)
+  }
+
+  private def convertEvents(version: Option[EventVersionPair]): Flow[WriteEvent, Event, NotUsed] = {
+    Flow[WriteEvent].statefulMapConcat(() => {
+      var currentVersion = version.getOrElse(EventVersionPair(seqNo = -1, txnNo = -1))
+      x => x match {
+        case WriteTransactionStart =>
+          val newVersion = currentVersion.copy(seqNo = currentVersion.seqNo + 1, txnNo = currentVersion.txnNo + 1)
```
Comment (cage433) [inline base/src/freetrm/eventstore/http/HttpWriter.scala]: use the increment methods?

File: base/src/freetrm/eventstore/http/HttpWriter.scala [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+  private val gzip = `Content-Encoding`(HttpEncodings.gzip)
+  private val headers = List(credentials, gzip)
+  
+  implicit val mat = ActorMaterializer.create(
+    ActorMaterializerSettings.create(system).withSyncProcessingLimit(1), system
+  )
+  implicit val dispatcher = system.dispatcher
+
+  def publish(topic: String, 
+              version: Option[EventVersionPair],
```
Comment (cage433) [inline base/src/freetrm/eventstore/http/HttpWriter.scala]: Moving `version` into `events`

File: base/tests/freetrm/eventstore/DeDuplicatorTest.scala [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+        events.contains(event) shouldBe false
+        events += event
+        Future.successful(event.version)
+      }
+      
+      override def close(): Unit = ???
+    }
+    
+    val dedup = new DeDuplicator(reader, writer)
+    val event = EventSourceEvent(EventVersionPair(1, 0), "key1", "hash", "data")
```
Comment (cage433) [inline base/tests/freetrm/eventstore/DeDuplicatorTest.scala]: Unused

File: base/tests/freetrm/eventstore/DeDuplicatorTest.scala [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+      """
+  )
+  
+  
+  test("de-duplicates") {
+    val reader = mock[EventSourceReader]
+    val topic = "topic"
+    val cv = "version"
+    (reader.topicExists _).expects(topic).returns(Future.successful(true)).once()
+    val topicMax = Some(10000l)
```
Comment (cage433) [inline base/tests/freetrm/eventstore/DeDuplicatorTest.scala]: This test is quite slow (> 3 s) - could reduce the numbers of events to 100/150

File: base/tests/freetrm/eventstore/DeDuplicatorTest.scala [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+      """,
+    "the SingleTopicDeDuplicator is created on demand and contains the right events" testsThat
+      """
+        When you write the first event for a topic the de-duplicator is loaded. Also tests that
+        the events loaded are correct.
+      """
+  )
+  
+  
+  test("de-duplicates") {
```
Comment (cage433) [inline base/tests/freetrm/eventstore/DeDuplicatorTest.scala:46]: This isn't testing that when we call with the same data as the latest surface, but _with a different version no_, it doesn't get written

File: base/tests/freetrm/eventstore/DeDuplicatorTest.scala [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+    (reader.consume _).expects(topic1, 0, topic1Max).returns(
+      messages
+    ).once()
+    (reader.consume _).expects(topic2, 0, topic2Max).returns(
+      messages
+    ).once()
+
+    val dedup = new DeDuplicator(reader, writer = null) // writer can be null as we're trying to write duplicates below
+    dedup.writeWithDuplicateCheckResult(topic1, 
+      EventSourceEvent(EventVersionPair(3, 0), "key1", "hash", "data")).futureValue shouldEqual
```
Comment (cage433) [inline base/tests/freetrm/eventstore/DeDuplicatorTest.scala]: should be version 4

File: base/src/freetrm/eventstore/http/EventStoreHttpServer.scala (hunk 9 above)
Comment (davidcorcoran) [inline base/src/freetrm/eventstore/http/EventStoreHttpServer.scala:56]: yup, `test("invalid get returns 404")` in `EventStoreHttpServiceTest`

File: base/tests/freetrm/eventstore/http/EventStoreHttpServiceTest.scala [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+    "invalid credentials fail" testsThat
+      """
+         We get an AuthenticationFailedRejection for bad creds
+      """,
+    "post writer" testsThat
+      """
+         Checks that when we POST an Event to the eventstore we get a corresponding write call on the underlying
+         writer.
+      """,
+    "can consume from event store" testsThat
```
Comment (cage433) [inline base/tests/freetrm/eventstore/http/EventStoreHttpServiceTest.scala:49]: Could perhaps do with a test that consume works when `lastOffsetToReadTo = None`

**Pull request merged**